src/mono/mono/metadata/metadata.c

```diff
@@ -1502,6 +1502,135 @@ mono_metadata_decode_row_col_raw (const MonoTableInfo *t, int idx, guint col)
 	return 0;
 }
 
+// This will inline into mono_metadata_decode_value_simd on targets that can't use
+//  the simd version of the algorithm.
+MONO_ALWAYS_INLINE static guint32
+decode_value_scalar (const guint8 *ptr, const guint8 **new_ptr)
+{
+	guint32 result;
+	guint8 b = *ptr;
+
+	if ((b & 0x80) == 0){
+		result = b;
+		++ptr;
+	} else if ((b & 0x40) == 0){
+		result = ((b & 0x3f) << 8 | ptr [1]);
+		ptr += 2;
+	} else if (b != 0xff) {
+		result = ((b & 0x1f) << 24) |
+			(ptr [1] << 16) |
+			(ptr [2] << 8) |
+			ptr [3];
+		ptr += 4;
+	}
+	else {
+		result = (ptr [1] << 24) | (ptr [2] << 16) | (ptr [3] << 8) | ptr [4];
+		ptr += 5;
+	}
+
+	if (new_ptr)
+		*new_ptr = ptr;
+
+	return result;
+}
+
+// This is meant to be wrapped by our existing decode_value functions, and inline into them.
+MONO_ALWAYS_INLINE guint32
+mono_metadata_decode_value_simd (const guint8 *ptr, const guint8 **new_ptr)
```

Comment:
in some cases, the callers of `mono_metadata_decode_blob_size` and `mono_metadata_decode_value` have already done a bounds check (for example the custom attribute parsing code does it). in other cases it's probably cheap to add a bounds check.  maybe we can have the simd version on other platforms for those cases?

Reply:
That would make sense, i'll look into how hard it would be to be certain we've bounds-checked. The cost of the bounds check may outweigh the benefit of the SIMD, though.